crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -491,6 +491,28 @@ impl SlidingSync {
 
         Span::current().record("pos", &pos);
 
+        // There is a non-negligible difference MSC3575 and MSC4186 in how
+        // the `e2ee` extension works. When the client sends a request with
+        // no `pos`:
+        //
+        // * MSC3575 returns all device lists updates since the last request from the
+        //   device that asked for device lists (this works similarly to to-device
+        //   message handling),
+        // * MSC4186 returns no device lists updates, as it only returns changes since
+        //   the provided `pos` (which is `null` in this case); this is in line with
+        //   sync v2.
+        //
+        // Therefore, with MSC4186, the device list cache must be marked as to be
+        // re-downloaded if the `since` token is `None`, otherwise it's easy to miss
+        // device lists updates that happened between the previous request and the new
+        // “initial” request.
+        #[cfg(feature = "e2e-encryption")]
+        if pos.is_none() && self.inner.version.is_native() && self.is_e2ee_enabled() {
+            let olm_machine = self.inner.client.olm_machine().await;
+            let olm_machine = olm_machine.as_ref().ok_or(Error::NoOlmMachine)?;
```

Comment:
Perhaps we should add a log that this is happening?

Reply:
Good idea, done.